{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CoDeepNEAT demo\n",
    "## CS081 project checkpoint demo\n",
    "### Harsha Uppli, Alan Zhao, Gabriel Meyer-Lee\n",
    "\n",
    "The following notebook demonstrates using CoDeepNEAT to solve CIFAR-10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "from keras.datasets import cifar10\n",
    "from keras.utils.np_utils import to_categorical\n",
    "from math import pi, floor\n",
    "from random import random\n",
    "from codeepneat import codeepneat, config, population, chromosome, genome, visualize\n",
    "import pickle\n",
    "import numpy as np\n",
    "import keras"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem: CIFAR10 data set\n",
    "\n",
    "Conveniently, it's also built into Keras, which our CoDeepNEAT imiplementation is built off of."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "data shapes\n",
      "  x train: (50000, 32, 32, 3)\n",
      "  x test: (10000, 32, 32, 3)\n",
      "  y train: (50000, 10)\n",
      "  y test: (10000, 10)\n"
     ]
    }
   ],
   "source": [
    "(x_train, y_train), (x_test, y_test) = cifar10.load_data()\n",
    "y_train = y_train[:,0]\n",
    "y_test = y_test[:,0]\n",
    "\n",
    "x_train = np.reshape(x_train, (x_train.shape[0], 32, 32, 3)).astype('float32') / 255\n",
    "x_test = np.reshape(x_test, (x_test.shape[0], 32, 32, 3)).astype('float32') / 255\n",
    "y_train = keras.utils.np_utils.to_categorical(y_train)\n",
    "y_test = keras.utils.np_utils.to_categorical(y_test)\n",
    "data = [x_train, y_train, x_test, y_test]\n",
    "print(\"data shapes\")\n",
    "print(\"  x train:\", x_train.shape)\n",
    "print(\"  x test:\", x_test.shape)\n",
    "print(\"  y train:\", y_train.shape)\n",
    "print(\"  y test:\", y_test.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Configuring NEAT\n",
    "\n",
    "Many of the options and inputs are still handled through the config file. The config file has been shortened considerably as many parameters have been eliminated, although many parameters have also been introduced which could be added."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting configCIFAR10\n"
     ]
    }
   ],
   "source": [
    "%%file configCIFAR10\n",
    "#--- parameters for the robot experiment ---#\n",
    "[phenotype]\n",
    "input_nodes         = 32,32,3\n",
    "output_nodes        = 10\n",
    "conv                = True\n",
    "\n",
    "[genetic]\n",
    "max_fitness_threshold = 1\n",
    "\n",
    "# Human reasoning\n",
    "pop_size              = 10\n",
    "prob_addconv          = 0.9\n",
    "prob_addlayer         = 0.1\n",
    "prob_addmodule        = 0.05\n",
    "elitism               = 1\n",
    "\n",
    "[genotype compatibility]\n",
    "compatibility_threshold = 3.0\n",
    "compatibility_change    = 0.0\n",
    "excess_coefficient      = 5.0\n",
    "disjoint_coefficient    = 3.0\n",
    "connection_coefficient  = 0.4\n",
    "size_coefficient        = 0.8\n",
    "\n",
    "[species]\n",
    "species_size        = 10\n",
    "survival_threshold  = 0.2\n",
    "old_threshold       = 30\n",
    "youth_threshold     = 10\n",
    "old_penalty         = 0.2\n",
    "youth_boost         = 1.2\n",
    "max_stagnation      = 15"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fitness\n",
    "\n",
    "For this demonstration we'll be using supervised learning to train the networks produced by CoDeepNEAT on CIFAR-10 and will use their accuracy after 5 epochs as our fitness. CIFAR-10, like MNIST, is a 10 category classification problem."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fitness(network, data):\n",
    "    network.fit(data[0], data[1],  epochs=5)\n",
    "    loss, acc = network.evaluate(data[2], data[3])\n",
    "    return acc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Evolution\n",
    "\n",
    "Evolution with CoDeepNEAT is slightly different than evolution with NEAT. The main difference is coevolution, where we have two separate populations with a hierarchical relationship evolving together."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def evolve(n):\n",
    "    config.load('configCIFAR10')\n",
    "    # Create 2 separate populations (size is now defined explicitly, but config file can still be used)\n",
    "    module_pop = population.Population(10, chromosome.ModuleChromo)\n",
    "    # As the top hierarchical level, the blueprint population needs to be able to see the module population\n",
    "    blueprint_pop = population.Population(5, chromosome.BlueprintChromo, module_pop)\n",
    "    # Most of the actual evolving is now handled outside of the population, by CoDeepNEAT\n",
    "    # Instead of requiring the user to overwrite the evaluation function, CoDeepNEAT evaluates the populations itself,\n",
    "    # it simply requires a fitness function for the networks it creates passed in as an argument.\n",
    "    codeepneat.epoch(n, blueprint_pop, module_pop, 10, fitness, data, save_best=True, name='CIFAR10')\n",
    "    # It will still stop if fitness surpasses the max_fitness_threshold in config file\n",
    "    # Plots the evolution of the best/average fitness\n",
    "    visualize.plot_stats(module_pop.stats, name=\"CIFAR10mod_\")\n",
    "    visualize.plot_stats(blueprint_pop.stats, name=\"CIFAR10bp_\")\n",
    "    # Visualizes speciation\n",
    "    #visualize.plot_species(module_pop.species_log, name=\"NMISTmod_\")\n",
    "    #visualize.plot_species(blueprint_pop.species_log, name=\"NMISTbp_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----Generation 0--------\n",
      "ID: 6 Species: 1\n",
      "Genes:\n",
      "\tModule ID: 16 Species ID:  3\n",
      "\tModule ID: 17 Species ID:  2\n",
      "ID: 7 Species: 2\n",
      "Genes:\n",
      "\tModule ID: 18 Species ID:  1\n",
      "\tModule ID: 19 Species ID:  1\n",
      "\tModule ID: 20 Species ID:  2\n",
      "ID: 8 Species: 3\n",
      "Genes:\n",
      "\tModule ID: 21 Species ID:  1\n",
      "\tModule ID: 22 Species ID:  1\n",
      "\tModule ID: 23 Species ID:  2\n",
      "\tModule ID: 24 Species ID:  1\n",
      "ID: 9 Species: 4\n",
      "Genes:\n",
      "\tModule ID: 25 Species ID:  1\n",
      "\tModule ID: 26 Species ID:  3\n",
      "\tModule ID: 27 Species ID:  2\n",
      "ID: 10 Species: 2\n",
      "Genes:\n",
      "\tModule ID: 28 Species ID:  1\n",
      "\tModule ID: 29 Species ID:  2\n",
      "\tModule ID: 30 Species ID:  1\n",
      "ID: 11 Species: 1\n",
      "Genes:\n",
      "\tLayer ID:  8 Type:   CONV Size:  256\n",
      "ID: 12 Species: 1\n",
      "Genes:\n",
      "\tLayer ID:  9 Type:   CONV Size:   64\n",
      "ID: 13 Species: 1\n",
      "Genes:\n",
      "\tLayer ID: 10 Type:   CONV Size:   32\n",
      "ID: 14 Species: 1\n",
      "Genes:\n",
      "\tLayer ID: 11 Type:   CONV Size:  128\n",
      "ID: 15 Species: 1\n",
      "Genes:\n",
      "\tLayer ID: 12 Type:   CONV Size:   32\n",
      "ID: 16 Species: 2\n",
      "Genes:\n",
      "\tLayer ID: 13 Type:   CONV Size:   32\n",
      "ID: 17 Species: 1\n",
      "Genes:\n",
      "\tLayer ID: 14 Type:   CONV Size:  128\n",
      "ID: 18 Species: 1\n",
      "Genes:\n",
      "\tLayer ID: 15 Type:   CONV Size:  128\n",
      "ID: 19 Species: 1\n",
      "Genes:\n",
      "\tLayer ID: 16 Type:   CONV Size:  128\n",
      "ID: 20 Species: 3\n",
      "Genes:\n",
      "\tLayer ID:  4 Type:  DENSE Size:   32\n",
      "Network 0\n",
      "Epoch 1/5\n",
      "50000/50000 [==============================] - 31s - loss: 1.6351 - acc: 0.4260    \n",
      "Epoch 2/5\n",
      "50000/50000 [==============================] - 31s - loss: 1.4728 - acc: 0.4892    \n",
      "Epoch 3/5\n",
      "50000/50000 [==============================] - 31s - loss: 1.4227 - acc: 0.5095    \n",
      "Epoch 4/5\n",
      "50000/50000 [==============================] - 31s - loss: 1.3894 - acc: 0.5228    \n",
      "Epoch 5/5\n",
      "50000/50000 [==============================] - 31s - loss: 1.3670 - acc: 0.5346    \n",
      " 9952/10000 [============================>.] - ETA: 0s\n",
      "Network 0 Fitness: 0.4747\n",
      "Network 1\n",
      "Epoch 1/5\n",
      "50000/50000 [==============================] - 31s - loss: 1.6342 - acc: 0.4255    \n",
      "Epoch 2/5\n",
      "50000/50000 [==============================] - 31s - loss: 1.4783 - acc: 0.4865    \n",
      "Epoch 3/5\n",
      "50000/50000 [==============================] - 31s - loss: 1.4285 - acc: 0.5079    \n",
      "Epoch 4/5\n",
      "50000/50000 [==============================] - 31s - loss: 1.3981 - acc: 0.5196    \n",
      "Epoch 5/5\n",
      "50000/50000 [==============================] - 31s - loss: 1.3751 - acc: 0.5286    \n",
      " 9952/10000 [============================>.] - ETA: 0s\n",
      "Network 1 Fitness: 0.4928\n",
      "Network 2\n",
      "Epoch 1/5\n",
      "50000/50000 [==============================] - 58s - loss: 1.6221 - acc: 0.4324    \n",
      "Epoch 2/5\n",
      "50000/50000 [==============================] - 58s - loss: 1.4621 - acc: 0.4938    \n",
      "Epoch 3/5\n",
      "50000/50000 [==============================] - 58s - loss: 1.4090 - acc: 0.5168    \n",
      "Epoch 4/5\n",
      "50000/50000 [==============================] - 58s - loss: 1.3709 - acc: 0.5323    \n",
      "Epoch 5/5\n",
      "50000/50000 [==============================] - 58s - loss: 1.3380 - acc: 0.5448    \n",
      "10000/10000 [==============================] - 6s     \n",
      "\n",
      "Network 2 Fitness: 0.5048\n",
      "Network 3\n",
      "Epoch 1/5\n",
      "50000/50000 [==============================] - 7s - loss: 1.9240 - acc: 0.2955     \n",
      "Epoch 2/5\n",
      "50000/50000 [==============================] - 7s - loss: 1.8015 - acc: 0.3473     \n",
      "Epoch 3/5\n",
      "50000/50000 [==============================] - 7s - loss: 1.7362 - acc: 0.3779     \n",
      "Epoch 4/5\n",
      "50000/50000 [==============================] - 7s - loss: 1.7025 - acc: 0.3887     \n",
      "Epoch 5/5\n",
      "50000/50000 [==============================] - 7s - loss: 1.6836 - acc: 0.4014     \n",
      " 9696/10000 [============================>.] - ETA: 0s\n",
      "Network 3 Fitness: 0.3757\n",
      "Network 4\n",
      "Epoch 1/5\n",
      "50000/50000 [==============================] - 31s - loss: 1.6251 - acc: 0.4307    \n",
      "Epoch 2/5\n",
      "50000/50000 [==============================] - 31s - loss: 1.4677 - acc: 0.4936    \n",
      "Epoch 3/5\n",
      "50000/50000 [==============================] - 31s - loss: 1.4179 - acc: 0.5129    \n",
      "Epoch 4/5\n",
      "50000/50000 [==============================] - 31s - loss: 1.3870 - acc: 0.5273    \n",
      "Epoch 5/5\n",
      "50000/50000 [==============================] - 31s - loss: 1.3584 - acc: 0.5357    \n",
      "10000/10000 [==============================] - 2s     \n",
      "\n",
      "Network 4 Fitness: 0.4875\n",
      "Network 5\n",
      "Epoch 1/5\n",
      "50000/50000 [==============================] - 31s - loss: 1.6248 - acc: 0.4288    \n",
      "Epoch 2/5\n",
      "50000/50000 [==============================] - 31s - loss: 1.4669 - acc: 0.4923    \n",
      "Epoch 3/5\n",
      "50000/50000 [==============================] - 31s - loss: 1.4173 - acc: 0.5114    \n",
      "Epoch 4/5\n",
      "50000/50000 [==============================] - 31s - loss: 1.3856 - acc: 0.5266    \n",
      "Epoch 5/5\n",
      "50000/50000 [==============================] - 31s - loss: 1.3581 - acc: 0.5360    \n",
      " 9856/10000 [============================>.] - ETA: 0s\n",
      "Network 5 Fitness: 0.475\n",
      "Network 6\n",
      "Epoch 1/5\n",
      "50000/50000 [==============================] - 48s - loss: 1.6737 - acc: 0.4020    \n",
      "Epoch 2/5\n",
      "50000/50000 [==============================] - 47s - loss: 1.4762 - acc: 0.4778    \n",
      "Epoch 3/5\n",
      "50000/50000 [==============================] - 48s - loss: 1.4116 - acc: 0.5005    \n",
      "Epoch 4/5\n",
      "50000/50000 [==============================] - 48s - loss: 1.3644 - acc: 0.5195    \n",
      "Epoch 5/5\n",
      "50000/50000 [==============================] - 48s - loss: 1.3273 - acc: 0.5316    \n",
      "10000/10000 [==============================] - 6s     \n",
      "\n",
      "Network 6 Fitness: 0.4939\n",
      "Network 7\n",
      "Epoch 1/5\n",
      "50000/50000 [==============================] - 7s - loss: 2.0862 - acc: 0.2080     \n",
      "Epoch 2/5\n",
      "50000/50000 [==============================] - 6s - loss: 2.0128 - acc: 0.2285     \n",
      "Epoch 3/5\n",
      "50000/50000 [==============================] - 7s - loss: 1.9927 - acc: 0.2350     \n",
      "Epoch 4/5\n",
      "50000/50000 [==============================] - 7s - loss: 1.9770 - acc: 0.2424     \n",
      "Epoch 5/5\n",
      "50000/50000 [==============================] - 7s - loss: 1.9672 - acc: 0.2469     \n",
      " 9504/10000 [===========================>..] - ETA: 0s\n",
      "Network 7 Fitness: 0.2463\n",
      "Network 8\n",
      "Epoch 1/5\n",
      "50000/50000 [==============================] - 31s - loss: 1.6335 - acc: 0.4277    \n",
      "Epoch 2/5\n",
      "50000/50000 [==============================] - 31s - loss: 1.4717 - acc: 0.4898    \n",
      "Epoch 3/5\n",
      "50000/50000 [==============================] - 31s - loss: 1.4263 - acc: 0.5082    \n",
      "Epoch 4/5\n",
      "50000/50000 [==============================] - 31s - loss: 1.3963 - acc: 0.5206    \n",
      "Epoch 5/5\n",
      "50000/50000 [==============================] - 31s - loss: 1.3727 - acc: 0.5300    \n",
      " 9952/10000 [============================>.] - ETA: 0s\n",
      "Network 8 Fitness: 0.4957\n",
      "Network 9\n",
      "Epoch 1/5\n",
      "50000/50000 [==============================] - 48s - loss: 1.6916 - acc: 0.3900    \n",
      "Epoch 2/5\n",
      "50000/50000 [==============================] - 47s - loss: 1.5251 - acc: 0.4547    \n",
      "Epoch 3/5\n",
      "50000/50000 [==============================] - 47s - loss: 1.4591 - acc: 0.4806    \n",
      "Epoch 4/5\n",
      "50000/50000 [==============================] - 48s - loss: 1.4132 - acc: 0.4975    \n",
      "Epoch 5/5\n",
      "50000/50000 [==============================] - 48s - loss: 1.3762 - acc: 0.5107    \n",
      "10000/10000 [==============================] - 6s     \n",
      "\n",
      "Network 9 Fitness: 0.4788\n",
      "-----Blueprints----------\n",
      "\n",
      " ****** Generation 0 ****** \n",
      "\n",
      "Population's average fitness: 0.49746 stdev: 0.12338\n",
      "Best fitness: 0.7 - size: 4 - species 3 - id 8\n",
      "Species length: 4 totalizing 5 individuals\n",
      "Species ID       : [1, 2, 3, 4]\n",
      "Each species size: [1, 2, 1, 1]\n",
      "Amount to spawn  : [1, 1, 2, 1]\n",
      "Species age      : [0, 0, 0, 0]\n",
      "Species no improv: [0, 0, 0, 0]\n",
      "-----Modules-----------\n",
      "\n",
      " ****** Generation 0 ****** \n",
      "\n",
      "Population's average fitness: 0.56017 stdev: 0.11712\n",
      "Best fitness: 0.7 - size: 1 - species 1 - id 11\n",
      "Species length: 3 totalizing 10 individuals\n",
      "Species ID       : [1, 2, 3]\n",
      "Each species size: [8, 1, 1]\n",
      "Amount to spawn  : [4, 3, 3]\n",
      "Species age      : [0, 0, 0]\n",
      "Species no improv: [0, 0, 0]\n"
     ]
    },
    {
     "ename": "AssertionError",
     "evalue": "IN: Layer ID: 17 Type:   CONV Size:   64 OUT: Layer ID: 10 Type:   CONV Size:   32",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAssertionError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-14-c3b6faae429d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mevolve\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m5\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-13-ce64137c0d10>\u001b[0m in \u001b[0;36mevolve\u001b[0;34m(n)\u001b[0m\n\u001b[1;32m      8\u001b[0m     \u001b[0;31m# Instead of requiring the user to overwrite the evaluation function, CoDeepNEAT evaluates the populations itself,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m     \u001b[0;31m# it simply requires a fitness function for the networks it creates passed in as an argument.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 10\u001b[0;31m     \u001b[0mcodeepneat\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mepoch\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mblueprint_pop\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmodule_pop\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m10\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfitness\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msave_best\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'CIFAR10'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     11\u001b[0m     \u001b[0;31m# It will still stop if fitness surpasses the max_fitness_threshold in config file\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m     \u001b[0;31m# Plots the evolution of the best/average fitness\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/cs81/project-huppili1-zzhao1/experiment/CDNEAT/codeepneat/codeepneat.py\u001b[0m in \u001b[0;36mepoch\u001b[0;34m(n, pop1, pop2, num_networks, f, data, save_best, name, report)\u001b[0m\n\u001b[1;32m     83\u001b[0m         \u001b[0mj\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpop1\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mepoch\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mg\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreport\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mreport\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msave_best\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     84\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"-----Modules-----------\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 85\u001b[0;31m         \u001b[0mk\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpop2\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mepoch\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mg\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreport\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mreport\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msave_best\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     86\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0msave_best\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     87\u001b[0m             \u001b[0mfilename\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"best_model_\"\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/cs81/project-huppili1-zzhao1/experiment/CDNEAT/codeepneat/population.py\u001b[0m in \u001b[0;36mepoch\u001b[0;34m(self, g, report, save_best, checkpoint_interval, checkpoint_generation, name)\u001b[0m\n\u001b[1;32m    375\u001b[0m         \u001b[0;31m# Spawning new population\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    376\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0ms\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__species\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 377\u001b[0;31m             \u001b[0mnew_population\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mextend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ms\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreproduce\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    378\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    379\u001b[0m         \u001b[0;31m# Remove stagnated species and its members\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/cs81/project-huppili1-zzhao1/experiment/CDNEAT/codeepneat/species.py\u001b[0m in \u001b[0;36mreproduce\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    137\u001b[0m                 \u001b[0;32massert\u001b[0m \u001b[0mparent1\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mspecies_id\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mparent2\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mspecies_id\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;31m \u001b[0m\u001b[0;31m\\\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    138\u001b[0m                        \u001b[0;34m\"Parents has different species id.\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 139\u001b[0;31m                 \u001b[0mchild\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparent1\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcrossover\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mparent2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    140\u001b[0m                 \u001b[0moffspring\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mchild\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmutate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    141\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/cs81/project-huppili1-zzhao1/experiment/CDNEAT/codeepneat/chromosome.py\u001b[0m in \u001b[0;36mcrossover\u001b[0;34m(self, other)\u001b[0m\n\u001b[1;32m     65\u001b[0m         \u001b[0mchild\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__class__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mid\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mother\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mid\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_gene_type\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     66\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 67\u001b[0;31m         \u001b[0mchild\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_inherit_genes\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mparent1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mparent2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     68\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     69\u001b[0m         \u001b[0mchild\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mspecies_id\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparent1\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mspecies_id\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/cs81/project-huppili1-zzhao1/experiment/CDNEAT/codeepneat/chromosome.py\u001b[0m in \u001b[0;36m_inherit_genes\u001b[0;34m(child, parent1, parent2)\u001b[0m\n\u001b[1;32m    311\u001b[0m                 \u001b[0;31m# copies disjoint genes from the fittest parent\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    312\u001b[0m                 \u001b[0mchild\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_genes\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mg1\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 313\u001b[0;31m         \u001b[0mparent1\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__connection_sort\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    314\u001b[0m         \u001b[0mchild\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_connections\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparent1\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_connections\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    315\u001b[0m         \u001b[0mchild\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__connection_sort\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/cs81/project-huppili1-zzhao1/experiment/CDNEAT/codeepneat/chromosome.py\u001b[0m in \u001b[0;36m__connection_sort\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    420\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    421\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__connection_sort\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 422\u001b[0;31m         \u001b[0;32massert\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_connections\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_in\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mid\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_connections\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    423\u001b[0m         \u001b[0;32massert\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_connections\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_out\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mid\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_connections\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    424\u001b[0m         \u001b[0mnew_conns\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAssertionError\u001b[0m: IN: Layer ID: 17 Type:   CONV Size:   64 OUT: Layer ID: 10 Type:   CONV Size:   32"
     ]
    }
   ],
   "source": [
    "evolve(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def eval_best(model_file):\n",
    "    config.load('configCIFAR10')\n",
    "    model = keras.models.load_model(model_file)\n",
    "    visualize.draw_net(model, \"_\" + model_file)    \n",
    "    model.fit(data[0], data[1], epochs=50)\n",
    "    loss, fitness = model.evaluate(data[2], data[3])\n",
    "    print(\"fitness\", fitness)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "eval_best(\"CIFAR10_best_model_0\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
