{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CoDeepNEAT demo\n",
    "## CS081 project checkpoint demo\n",
    "### Harsha Uppli, Alan Zhao, Gabriel Meyer-Lee\n",
    "\n",
    "The following notebook demonstrates using CoDeepNEAT to solve the Penn Tree Bank"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "from keras.datasets import reuters\n",
    "from keras.preprocessing.text import Tokenizer\n",
    "from keras.utils.np_utils import to_categorical\n",
    "from keras import preprocessing\n",
    "from keras import backend as K\n",
    "from math import pi, floor\n",
    "from random import random\n",
    "from codeepneat import codeepneat, config, population, chromosome, genome, visualize\n",
    "import pickle\n",
    "import numpy as np\n",
    "import keras\n",
    "import collections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading data...\n",
      "8982 train sequences\n",
      "2246 test sequences\n",
      "46 classes\n",
      "Vectorizing sequence data...\n",
      "Convert class vector to binary class matrix (for use with categorical_crossentropy)\n",
      "x_train shape (7840, 30)\n",
      "x_val shape (1120, 30)\n",
      "x_test shape (2240, 30)\n"
     ]
    }
   ],
   "source": [
    "max_words = 10000\n",
    "print('Loading data...')\n",
    "(x_train_all, y_train_all), (x_test, y_test) = reuters.load_data(num_words=max_words,\n",
    "                                                         test_split=0.2)\n",
    "print(len(x_train_all), 'train sequences')\n",
    "print(len(x_test), 'test sequences')\n",
    "\n",
    "num_classes = np.max(y_train_all) + 1\n",
    "print(num_classes, 'classes')\n",
    "\n",
    "print('Vectorizing sequence data...')\n",
    "tokenizer = Tokenizer(num_words=max_words)\n",
    "x_train_all = tokenizer.sequences_to_matrix(x_train_all, mode='binary')\n",
    "x_test = tokenizer.sequences_to_matrix(x_test, mode='binary')\n",
    "x_train_all = preprocessing.sequence.pad_sequences(x_train_all, maxlen=30)\n",
    "x_test = preprocessing.sequence.pad_sequences(x_test, maxlen=30)[:2240]\n",
    "\n",
    "# print('x_train shape:', x_train.shape)\n",
    "# print('x_test shape:', x_test.shape)\n",
    "\n",
    "\n",
    "print('Convert class vector to binary class matrix '\n",
    "      '(for use with categorical_crossentropy)')\n",
    "y_train_all = keras.utils.to_categorical(y_train_all, num_classes)\n",
    "y_test = keras.utils.to_categorical(y_test, num_classes)[:2240]\n",
    "#Reshaping the input data\n",
    "# x_train = np.reshape(x_train, (8982, 10000, 1))\n",
    "# x_test = np.reshape(x_test, (2246, 10000, 1))\n",
    "# print('y_train shape:', y_train_all.shape)\n",
    "# print('y_test shape:', y_test.shape)\n",
    "\n",
    "index = np.array(range(len(x_train_all)))\n",
    "np.random.shuffle(index)\n",
    "index_train = index[:7862]\n",
    "index_val = index[7862:]\n",
    "x_train = x_train_all[index_train][:7840]\n",
    "y_train = y_train_all[index_train][:7840]\n",
    "x_val = x_train_all[index_val]\n",
    "y_val = y_train_all[index_val]\n",
    "print('x_train shape', x_train.shape)\n",
    "print('x_val shape', x_val.shape)\n",
    "print('x_test shape', x_test.shape)\n",
    "\n",
    "data = [x_train, y_train, x_val, y_val, x_test, y_test]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting configReuters\n"
     ]
    }
   ],
   "source": [
    "%%file configReuters\n",
    "#--- parameters for the robot experiment ---#\n",
    "[phenotype]\n",
    "input_nodes         = 30\n",
    "output_nodes        = 46\n",
    "conv                = False\n",
    "LSTM                = True\n",
    "\n",
    "[genetic]\n",
    "max_fitness_threshold = 1\n",
    "\n",
    "# Human reasoning\n",
    "pop_size              = 10\n",
    "prob_addconv          = 0.0\n",
    "prob_addLSTM          = 0.9\n",
    "prob_addlayer         = 0.1\n",
    "prob_mutatelayer      = 0.4\n",
    "prob_addmodule        = 0.05\n",
    "prob_switchmodule     = 0.3\n",
    "elitism               = 1\n",
    "\n",
    "[genotype compatibility]\n",
    "compatibility_threshold = 3.0\n",
    "compatibility_change    = 0.0\n",
    "excess_coefficient      = 5.0\n",
    "disjoint_coefficient    = 3.0\n",
    "connection_coefficient  = 0.4\n",
    "size_coefficient        = 0.8\n",
    "\n",
    "[species]\n",
    "species_size        = 10\n",
    "survival_threshold  = 0.2\n",
    "old_threshold       = 30\n",
    "youth_threshold     = 10\n",
    "old_penalty         = 0.2\n",
    "youth_boost         = 1.2\n",
    "max_stagnation      = 15"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fitness(network, data):\n",
    "    K.set_learning_phase(1)\n",
    "    network.fit(data[0], data[1],  epochs=5, batch_size = 32)\n",
    "    K.set_learning_phase(0)\n",
    "    loss, acc = network.evaluate(data[2], data[3], batch_size=32)\n",
    "    return acc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def evolve(n, debugging=False):\n",
    "    if(debugging):\n",
    "        debug = open(\"debug.txt\", \"w\")\n",
    "    else:\n",
    "        debug = None\n",
    "    config.load('configReuters')\n",
    "    # Create 2 separate populations (size is now defined explicitly, but config file can still be used)\n",
    "    module_pop = population.Population(15, chromosome.ModuleChromo, debug=debug)\n",
    "    # As the top hierarchical level, the blueprint population needs to be able to see the module population\n",
    "    blueprint_pop = population.Population(10, chromosome.BlueprintChromo, module_pop, debug=debug)\n",
    "    # Most of the actual evolving is now handled outside of the population, by CoDeepNEAT\n",
    "    # Instead of requiring the user to overwrite the evaluation function, CoDeepNEAT evaluates the populations itself,\n",
    "    # it simply requires a fitness function for the networks it creates passed in as an argument.\n",
    "    codeepneat.epoch(n, blueprint_pop, module_pop, 2, fitness, data, save_best=True, name='reuters', debug=debug)\n",
    "    # It will still stop if fitness surpasses the max_fitness_threshold in config file\n",
    "    # Plots the evolution of the best/average fitness\n",
    "    visualize.plot_stats(module_pop.stats, name=\"reutersmod_\")\n",
    "    visualize.plot_stats(blueprint_pop.stats, name=\"reutersbp_\")\n",
    "    # Visualizes speciation\n",
    "    #visualize.plot_species(module_pop.species_log, name=\"NMISTmod_\")\n",
    "    #visualize.plot_species(blueprint_pop.species_log, name=\"NMISTbp_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----Generation 0--------\n",
      "Network 0\n",
      "Epoch 1/5\n",
      "7840/7840 [==============================] - 25s - loss: 2.4623 - acc: 0.3417    \n",
      "Epoch 2/5\n",
      "7840/7840 [==============================] - 25s - loss: 2.4330 - acc: 0.3494    \n",
      "Epoch 3/5\n",
      "7840/7840 [==============================] - 24s - loss: 2.4289 - acc: 0.3494    \n",
      "Epoch 4/5\n",
      "7840/7840 [==============================] - 24s - loss: 2.4270 - acc: 0.3494    \n",
      "Epoch 5/5\n",
      "7840/7840 [==============================] - 24s - loss: 2.4257 - acc: 0.3494    \n",
      "1056/1120 [===========================>..] - ETA: 0s\n",
      "Network 0 Fitness: 0.366964285714\n",
      "Network 1\n",
      "Epoch 1/5\n",
      "7840/7840 [==============================] - 25s - loss: 2.5196 - acc: 0.3401    \n",
      "Epoch 2/5\n",
      "7840/7840 [==============================] - 25s - loss: 2.4278 - acc: 0.3494    \n",
      "Epoch 3/5\n",
      "7840/7840 [==============================] - 24s - loss: 2.4258 - acc: 0.3494    \n",
      "Epoch 4/5\n",
      "7840/7840 [==============================] - 24s - loss: 2.4239 - acc: 0.3494    \n",
      "Epoch 5/5\n",
      "7840/7840 [==============================] - 23s - loss: 2.4339 - acc: 0.3467    \n",
      "1056/1120 [===========================>..] - ETA: 0s\n",
      "Network 1 Fitness: 0.366964285714\n",
      "-----Modules-----------\n",
      "\n",
      " ****** Generation 0 ****** \n",
      "\n",
      "Population's average fitness: 0.37563 stdev: 0.00340\n",
      "Best fitness: 0.3769642857 - size: 1 - species 1 - id 16\n",
      "Species length: 4 totalizing 15 individuals\n",
      "Species ID       : [1, 2, 3, 4]\n",
      "Each species size: [12, 1, 1, 1]\n",
      "Amount to spawn  : [4, 4, 4, 4]\n",
      "Species age      : [0, 0, 0, 0]\n",
      "Species no improv: [0, 0, 0, 0]\n",
      "Removing 1 excess indiv(s) from the new population\n",
      "-----Blueprints----------\n",
      "\n",
      " ****** Generation 0 ****** \n",
      "\n",
      "Population's average fitness: 0.37496 stdev: 0.00400\n",
      "Best fitness: 0.3769642857 - size: 4 - species 1 - id 11\n",
      "Species length: 4 totalizing 10 individuals\n",
      "Species ID       : [1, 2, 3, 4]\n",
      "Each species size: [3, 2, 1, 4]\n",
      "Amount to spawn  : [3, 3, 3, 2]\n",
      "Species age      : [0, 0, 0, 0]\n",
      "Species no improv: [0, 0, 0, 0]\n",
      "Removing 1 excess indiv(s) from the new population\n",
      "-----Generation 1--------\n",
      "Network 0\n",
      "Epoch 1/5\n",
      "7840/7840 [==============================] - 31s - loss: 2.5034 - acc: 0.3342    \n",
      "Epoch 2/5\n",
      "7840/7840 [==============================] - 31s - loss: 2.4367 - acc: 0.3485    \n",
      "Epoch 3/5\n",
      "7840/7840 [==============================] - 32s - loss: 2.4313 - acc: 0.3491    \n",
      "Epoch 4/5\n",
      "7840/7840 [==============================] - 32s - loss: 2.4332 - acc: 0.3494    \n",
      "Epoch 5/5\n",
      "7840/7840 [==============================] - 32s - loss: 2.4375 - acc: 0.3473    \n",
      "1120/1120 [==============================] - 1s     \n",
      "\n",
      "Network 0 Fitness: 0.366964285714\n",
      "Network 1\n",
      "Epoch 1/5\n",
      "7840/7840 [==============================] - 1s - loss: 2.7009 - acc: 0.3480     \n",
      "Epoch 2/5\n",
      "7840/7840 [==============================] - 1s - loss: 2.7578 - acc: 0.3142     \n",
      "Epoch 3/5\n",
      "7840/7840 [==============================] - 1s - loss: 3.5647 - acc: 0.2788     \n",
      "Epoch 4/5\n",
      "7840/7840 [==============================] - 1s - loss: 3.3055 - acc: 0.2990     \n",
      "Epoch 5/5\n",
      "7840/7840 [==============================] - 1s - loss: 2.8639 - acc: 0.3151     \n",
      "  32/1120 [..............................] - ETA: 11s\n",
      "Network 1 Fitness: 0.366964285714\n",
      "-----Modules-----------\n",
      "\n",
      " ****** Generation 1 ****** \n",
      "\n",
      "Population's average fitness: 0.37563 stdev: 0.00340\n",
      "Best fitness: 0.3769642857 - size: 1 - species 1 - id 29\n",
      "Species length: 2 totalizing 15 individuals\n",
      "Species ID       : [1, 2]\n",
      "Each species size: [7, 8]\n",
      "Amount to spawn  : [7, 8]\n",
      "Species age      : [1, 1]\n",
      "Species no improv: [0, 1]\n",
      "-----Blueprints----------\n",
      "\n",
      " ****** Generation 1 ****** \n",
      "\n",
      "Population's average fitness: 0.37496 stdev: 0.00400\n",
      "Best fitness: 0.3769642857 - size: 4 - species 1 - id 18\n",
      "Species length: 4 totalizing 10 individuals\n",
      "Species ID       : [1, 2, 3, 4]\n",
      "Each species size: [3, 3, 3, 1]\n",
      "Amount to spawn  : [3, 3, 2, 2]\n",
      "Species age      : [1, 1, 1, 1]\n",
      "Species no improv: [1, 1, 1, 1]\n"
     ]
    },
    {
     "ename": "RuntimeError",
     "evalue": "Can't decrement id ref count (file write failed: time = Sun Dec  3 21:27:43 2017\n, filename = 'reuters_best_model_1', file descriptor = 72, errno = 122, error message = 'Disk quota exceeded', buf = 0x12dc1338, total write size = 8192, bytes this sub-write = 8192, bytes actually written = 18446744073709551615, offset = 19262616)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "\u001b[0;32mh5py/_objects.pyx\u001b[0m in \u001b[0;36mh5py._objects.ObjectID.__dealloc__\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mRuntimeError\u001b[0m: Can't decrement id ref count (file write failed: time = Sun Dec  3 21:27:43 2017\n, filename = 'reuters_best_model_1', file descriptor = 72, errno = 122, error message = 'Disk quota exceeded', buf = 0x12dc1338, total write size = 8192, bytes this sub-write = 8192, bytes actually written = 18446744073709551615, offset = 19262616)"
     ]
    }
   ],
   "source": [
    "evolve(25, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
